fix #30480 torch.normal shape checking is broken (#32243)

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33050 fix #30480 torch.normal shape checking is broken (#32243)**

Summary:
Following what @gchanan proposed in #30480

If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.

Also adding backward compatible for ATen cpu version normal(). Cpu version normal() is migrated from TH to ATen recently, but lack of backward compatible support.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

Commit:
fix #30480 torch.normal shape checking is broken (#32243)

Summary:
Following what gchanan proposed in #30480
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken (#32243)"


Summary:
Following what gchanan proposed in #30480
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken (#32243)"


Summary:
Following what gchanan proposed in #30480
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken (#32243)"


Summary:
normal CPU version has been moved from TH to aten already in the following pull request.
https://github.com/pytorch/pytorch/pull/32031
We no longer need to respect the old TH behavior.

Based on what @gchanan proposed in #30480, only broadcast should be supported now. The behavior of this code is as below:
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
- If the (logical) shapes of mean and std are not broadcastable, error out.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/Distributions.cpp:
```diff
@@ -234,8 +234,43 @@ Tensor& normal_out_cpu(Tensor& output, double mean, const Tensor& std, Generator
+  bool expandable = are_expandable(mean.sizes(), std.sizes());
+  bool empty_output = output.numel() == 0;
+
+  if (expandable) {
```
A lot of this logic is duplicated among normal_out_cpu and normal_out_cuda. Could we write a helper function to deduplicate the logic?

Something like
```
bool is_legacy_impl = resizeOutputForNormal(output, mean, std);
normal_cpu_(output, 0, 1, gen);
if (is_legacy_impl) {
  output.mul_(std.reshape(mean.sizes())).add_(mean);
}
else {
  output.mul_(std).add_(mean);
}
```

glaringlee replied:
done

zou3519 commented on aten/src/ATen/native/Distributions.cpp:
```diff
@@ -234,8 +234,43 @@ Tensor& normal_out_cpu(Tensor& output, double mean, const Tensor& std, Generator
+        "inconsistent tensor, std and mean are not broadcastable, output size (", output.sizes(), ") is not the same as mean size (", mean.sizes(), ")");
+    TORCH_WARN_ONCE(
+        "std and mean have the same number of elements, but are not broadcastable. This was previously a "
+        "supported mode of operation, but is now deprecated and the support will be removed in a later release. "
```
We should probably state a release number (1.6) and remember to file an issue for this after this PR gets merged

glaringlee replied:
done

zou3519 commented on test/test_torch.py:
```diff
@@ -5307,6 +5307,52 @@ def test_tensor_grad_warnings(self):
+
+            # inputs are non-expandable tensors, but they have same number of elements
+            self.assertEqual(torch.normal(tensor120, tensor2345).size(), (120,))
+            self.assertEqual(torch.normal(tensor2345, tensor120).size(), (2, 3, 4, 5))
```
If possible, we should also check that in this case (inputs have same number of elements) the deprecation warning gets raised. I am not sure what the semantics of TORCH_WARN_ONCE is so I am not sure how plausible this suggestion is

glaringlee replied:
Curerntly we have no way to catch this since TORCH_WARN_ONCE is not an assertion. But this one is covered in test case's log. It will be printed every time the test case is run.

zou3519 replied:
You should be able to catch python warnings using the python warnings package, as long as the function consistently raises a warning.  For example: https://github.com/pytorch/pytorch/blob/af4d6120bda2cb930b9b51b69a776ebbb9203616/test/test_torch.py#L1582-L1585

glaringlee replied:
Resolved by using assertWarnsRegex

zou3519 reviewed (approved):
We should try to share code between Distributions.cpp and Distributions.cu. Also, it would be nice if we could test that the deprecation warning actually shows up

Commit:
Update on "fix #30480 torch.normal shape checking is broken (#32243)"


Summary:
Following what @gchanan proposed in #30480

If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.

Also adding backward compatible for ATen cpu version normal(). Cpu version normal() is migrated from TH to ATen recently, but lack of backward compatible support.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken (#32243)"


Summary:
Following what @gchanan proposed in #30480

If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.

Also adding backward compatible for ATen cpu version normal(). Cpu version normal() is migrated from TH to ATen recently, but lack of backward compatible support.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken (#32243)"


Summary:
Following what @gchanan proposed in #30480

If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.

Also adding backward compatible for ATen cpu version normal(). Cpu version normal() is migrated from TH to ATen recently, but lack of backward compatible support.

Differential Revision: [D19771186](https://our.internmc.facebook.com/intern/diff/D19771186)

[ghstack-poisoned]
